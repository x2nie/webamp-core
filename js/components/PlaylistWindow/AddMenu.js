import React from "react";
import PlaylistMenu from "./PlaylistMenu";

/* eslint-disable no-alert */

const AddMenu = ({ openFileDialog }) => (
  <PlaylistMenu id="playlist-add-menu">
    <div
      className="add-url"
      onClick={() => alert("Not supported in Winamp2-js")}
    />
    <div
      className="add-dir"
      onClick={() => alert("Not supported in Winamp2-js")}
    />
    <div className="add-file" onClick={openFileDialog} />
  </PlaylistMenu>
);

export default AddMenu;
