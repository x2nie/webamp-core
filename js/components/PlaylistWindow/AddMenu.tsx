import React from "react";
import * as Selectors from "../../selectors";
import * as Actions from "../../actionCreators";
import PlaylistMenu from "./PlaylistMenu";

import { useTypedSelector, useActionCreator } from "../../hooks";

const AddMenu = () => {
  const nextIndex = useTypedSelector(Selectors.getTrackCount);
  const addDirAtIndex = useActionCreator(Actions.addDirAtIndex);
  const addFilesAtIndex = useActionCreator(Actions.addFilesAtIndex);
  return (
    <PlaylistMenu id="playlist-add-menu">
      <div
        className="add-url"
        onClick={() => window.alert("Not supported in Webamp")}
      />
      <div className="add-dir" onClick={() => addDirAtIndex(nextIndex)} />
      <div className="add-file" onClick={() => addFilesAtIndex(nextIndex)} />
    </PlaylistMenu>
  );
};

export default AddMenu;
