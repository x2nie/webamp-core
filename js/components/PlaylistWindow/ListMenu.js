import React from "react";
import { connect } from "react-redux";
import PlaylistMenu from "./PlaylistMenu";

/* eslint-disable no-alert */

const ListMenu = () => (
  <PlaylistMenu id="playlist-list-menu">
    <div
      className="new-list"
      onClick={() => alert("Not supported in Winamp2-js")}
    />
    <div
      className="save-list"
      onClick={() => alert("Not supported in Winamp2-js")}
    />
    <div
      className="load-list"
      onClick={() => alert("Not supported in Winamp2-js")}
    />
  </PlaylistMenu>
);

const mapDispatchToProps = {};
export default connect(null, mapDispatchToProps)(ListMenu);
