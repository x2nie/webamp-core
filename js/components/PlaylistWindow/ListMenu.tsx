import React from "react";
import { connect } from "react-redux";
import * as Actions from "../../actionCreators";
import PlaylistMenu from "./PlaylistMenu";
import { Dispatch } from "../../types";

/* eslint-disable no-alert */

interface DispatchProps {
  removeAllTracks: () => void;
}

const ListMenu = (props: DispatchProps) => (
  <PlaylistMenu id="playlist-list-menu">
    <div className="new-list" onClick={props.removeAllTracks} />
    <div
      className="save-list"
      onClick={() => alert("Not supported in Webamp")}
    />
    <div
      className="load-list"
      onClick={() => alert("Not supported in Webamp")}
    />
  </PlaylistMenu>
);

const mapDispatchToProps = (dispatch: Dispatch): DispatchProps => ({
  removeAllTracks: () => dispatch(Actions.removeAllTracks()),
});

export default connect(
  null,
  mapDispatchToProps
)(ListMenu);
