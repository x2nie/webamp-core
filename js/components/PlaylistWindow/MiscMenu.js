import React from "react";
import { connect } from "react-redux";
import {
  reverseList,
  randomizeList,
  sortListByTitle
} from "../../actionCreators";
import PlaylistMenu from "./PlaylistMenu";
import { ContextMenu, Hr, Node } from "../ContextMenu";

/* eslint-disable no-alert */
/* TODO: This should really be kitty-corner to the upper right hand corner of the MiscMenu */
const SortContextMenu = props => (
  <ContextMenu style={{ width: "100%", height: "100%" }} top handle={<div />}>
    <Node label="Sort list by title" onClick={props.sortListByTitle} />
    <Hr />
    <Node label="Reverse list" onClick={props.reverseList} />
    <Node label="Randomize list" onClick={props.randomizeList} />
  </ContextMenu>
);

const ConnectedSortcontextMenu = connect(null, {
  reverseList,
  randomizeList,
  sortListByTitle
})(SortContextMenu);

const MiscMenu = () => (
  <PlaylistMenu id="playlist-misc-menu">
    <div className="sort-list" onClick={e => e.stopPropagation()}>
      <ConnectedSortcontextMenu />
    </div>
    <div
      className="file-info"
      onClick={() => alert("Not supported in Winamp2-js")}
    />
    <div
      className="misc-options"
      onClick={() => alert("Not supported in Winamp2-js")}
    />
  </PlaylistMenu>
);

export default MiscMenu;
