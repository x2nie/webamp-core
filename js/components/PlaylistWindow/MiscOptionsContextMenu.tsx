import React from "react";
import { connect } from "react-redux";
import { Node } from "../ContextMenu";
import ContextMenuTarget from "../ContextMenuTarget";
import { Dispatch } from "../../types";
import { downloadHtmlPlaylist } from "../../actionCreators";

interface DispatchProps {
  downloadHtmlPlaylist: () => void;
}

const MiscOptionsContextMenu = (props: DispatchProps) => (
  <ContextMenuTarget
    top
    renderMenu={() => (
      <Node
        onClick={props.downloadHtmlPlaylist}
        label="Generate HTML playlist"
      />
    )}
  >
    <div />
  </ContextMenuTarget>
);

const mapDispatchToProps = (dispatch: Dispatch): DispatchProps => {
  return {
    downloadHtmlPlaylist: () => dispatch(downloadHtmlPlaylist()),
  };
};
export const ConnectedMiscOptionsContextMenu = connect(
  null,
  mapDispatchToProps
)(MiscOptionsContextMenu);
