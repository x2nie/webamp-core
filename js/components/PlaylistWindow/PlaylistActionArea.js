import React from "react";
import { connect } from "react-redux";

import {
  play,
  pause,
  stop,
  next,
  previous,
  openFileDialog
} from "../../actionCreators";

import MiniTime from "../MiniTime";
import RunningTimeDisplay from "./RunningTimeDisplay";

const PlaylistWindow = props => (
  <React.Fragment>
    <RunningTimeDisplay />
    <div className="playlist-action-buttons">
      <div className="playlist-previous-button" onClick={props.previous} />
      <div className="playlist-play-button" onClick={props.play} />
      <div className="playlist-pause-button" onClick={props.pause} />
      <div className="playlist-stop-button" onClick={props.stop} />
      <div className="playlist-next-button" onClick={props.next} />
      <div className="playlist-eject-button" onClick={props.openFileDialog} />
    </div>
    <MiniTime />
  </React.Fragment>
);

const mapDispatchToProps = {
  play,
  pause,
  stop,
  openFileDialog,
  next,
  previous
};

export default connect(null, mapDispatchToProps)(PlaylistWindow);
