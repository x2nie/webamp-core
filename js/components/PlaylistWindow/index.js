import React from "react";
import { connect } from "react-redux";
import classnames from "classnames";
import Slider from "rc-slider/lib/Slider";

import DropTarget from "../DropTarget";
import MiniTime from "../MiniTime";
import PlaylistShade from "./PlaylistShade";
import Track from "./Track";
import AddMenu from "./AddMenu";
import RemoveMenu from "./RemoveMenu";
import SelectionMenu from "./SelectionMenu";
import MiscMenu from "./MiscMenu";
import ListMenu from "./ListMenu";
import ResizeTarget from "./ResizeTarget";
import RunningTimeDisplay from "./RunningTimeDisplay";
import { percentToIndex } from "../../utils";
import {
  WINDOWS,
  PLAYLIST_RESIZE_SEGMENT_WIDTH,
  PLAYLIST_RESIZE_SEGMENT_HEIGHT,
  MIN_PLAYLIST_WINDOW_WIDTH
} from "../../constants";
import {
  TOGGLE_PLAYLIST_WINDOW,
  TOGGLE_PLAYLIST_SHADE_MODE,
  SET_FOCUSED_WINDOW,
  SET_PLAYLIST_SCROLL_POSITION
} from "../../actionTypes";
import { getOrderedTracks } from "../../selectors";
import {
  play,
  pause,
  stop,
  openFileDialog,
  toggleVisualizerStyle
} from "../../actionCreators";

import "../../../css/playlist-window.css";

const TRACK_HEIGHT = 13;
const MIN_WINDOW_HEIGHT = 116;

const Handle = () => <div className="playlist-scrollbar-handle" />;

const PlaylistWindow = props => {
  const {
    skinPlaylistStyle,
    focusPlaylist,
    focused,
    playlistScrollPosition,
    setPlaylistScrollPosition,
    trackOrder,
    playlistSize,
    playlistShade,
    close,
    toggleShade
  } = props;
  if (playlistShade) {
    return <PlaylistShade />;
  }

  const style = {
    color: skinPlaylistStyle.normal,
    backgroundColor: skinPlaylistStyle.normalbg,
    fontFamily: `${skinPlaylistStyle.font}, Arial, sans-serif`,
    height: `${MIN_WINDOW_HEIGHT +
      playlistSize[1] * PLAYLIST_RESIZE_SEGMENT_HEIGHT}px`,
    width: `${MIN_PLAYLIST_WINDOW_WIDTH +
      playlistSize[0] * PLAYLIST_RESIZE_SEGMENT_WIDTH}px`
  };

  const classes = classnames("window", "draggable", {
    selected: focused === WINDOWS.PLAYLIST,
    wide: playlistSize[0] > 2
  });

  const BASE_WINDOW_HEIGHT = 52;
  const numberOfVisibleTracks = Math.floor(
    (BASE_WINDOW_HEIGHT + PLAYLIST_RESIZE_SEGMENT_HEIGHT * playlistSize[1]) /
      TRACK_HEIGHT
  );
  const overflowTracks = Math.max(0, trackOrder.length - numberOfVisibleTracks);
  const offset = percentToIndex(
    playlistScrollPosition / 100,
    overflowTracks + 1
  );

  // Ugh. By not rendering some tracks, we can end up in a situation where
  // scrolling causes the number of digits in the tracks to go up, thus causing
  // a horizontal jump.
  const tracks = trackOrder.slice(offset, offset + numberOfVisibleTracks);

  return (
    <DropTarget
      id="playlist-window"
      className={classes}
      style={style}
      onMouseDown={focusPlaylist}
    >
      <div className="playlist-top draggable">
        <div className="playlist-top-left draggable" />
        <div className="playlist-top-title draggable" />
        <div className="playlist-top-right draggable">
          <div id="playlist-shade-button" onClick={toggleShade} />
          <div id="playlist-close-button" onClick={close} />
        </div>
      </div>
      <div className="playlist-middle draggable">
        <div className="playlist-middle-left draggable" />
        <div className="playlist-middle-center">
          <div className="playlist-tracks">
            <div>
              {tracks.map((id, i) => (
                <Track number={i + 1 + offset} id={id} key={id} />
              ))}
            </div>
          </div>
        </div>
        <div className="playlist-middle-right draggable">
          <Slider
            className="playlist-scrollbar"
            type="range"
            min={0}
            max={100}
            step={1}
            value={playlistScrollPosition}
            onChange={setPlaylistScrollPosition}
            vertical
            handle={Handle}
            disabled={overflowTracks === 0}
          />
        </div>
      </div>
      <div className="playlist-bottom draggable">
        <div className="playlist-bottom-left draggable">
          <AddMenu openFileDialog={props.openFileDialog} />
          <RemoveMenu />
          <SelectionMenu />
          <MiscMenu />
        </div>
        <div className="playlist-bottom-center draggable" />
        <div className="playlist-bottom-right draggable">
          <div
            className="playlist-visualizer"
            onClick={props.toggleVisualizerStyle}
          />
          <RunningTimeDisplay />
          <div className="playlist-action-buttons">
            <div className="playlist-previous-button" />
            <div className="playlist-play-button" onClick={props.play} />
            <div className="playlist-pause-button" onClick={props.pause} />
            <div className="playlist-stop-button" onClick={props.stop} />
            <div className="playlist-next-button" />
            <div
              className="playlist-eject-button"
              onClick={props.openFileDialog}
            />
          </div>
          <MiniTime />
          <ListMenu />
          <ResizeTarget />
        </div>
      </div>
    </DropTarget>
  );
};

const mapDispatchToProps = (dispatch, ownProps) => ({
  focusPlaylist: () =>
    dispatch({
      type: SET_FOCUSED_WINDOW,
      window: WINDOWS.PLAYLIST
    }),
  play: () => dispatch(play()),
  pause: () => dispatch(pause()),
  stop: () => dispatch(stop()),
  openFileDialog: () => dispatch(openFileDialog(ownProps.fileInput)),
  setPlaylistScrollPosition: position =>
    dispatch({ type: SET_PLAYLIST_SCROLL_POSITION, position: 100 - position }),
  close: () => dispatch({ type: TOGGLE_PLAYLIST_WINDOW }),
  toggleShade: () => dispatch({ type: TOGGLE_PLAYLIST_SHADE_MODE }),
  toggleVisualizerStyle: () => dispatch(toggleVisualizerStyle())
});

const mapStateToProps = state => {
  const {
    windows: { focused },
    display: {
      skinPlaylistStyle,
      playlistScrollPosition,
      playlistSize,
      playlistShade
    },
    media: { duration }
  } = state;
  return {
    focused,
    skinPlaylistStyle,
    playlistScrollPosition,
    playlistSize,
    playlistShade,
    trackOrder: getOrderedTracks(state),
    duration
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(PlaylistWindow);
