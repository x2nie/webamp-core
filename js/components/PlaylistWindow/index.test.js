const Winamp = {
  media: {
    addEventListener: jest.fn(),
    _analyser: null
  }
};

window.requestAnimationFrame = () => {};

import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import getStore from "../../store";

import PlaylistWindow from "./index";

describe("PlaylistWindow", () => {
  let store;
  beforeEach(() => {
    store = getStore(Winamp);
  });

  it("renders to snapshot", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <PlaylistWindow />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
