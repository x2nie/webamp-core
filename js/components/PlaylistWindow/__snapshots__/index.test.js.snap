// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaylistWindow renders to snapshot 1`] = `
<div
  className="window draggable"
  id="playlist-window"
  onDragEnter={[Function]}
  onDragOver={[Function]}
  onDrop={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "backgroundColor": undefined,
      "color": undefined,
      "fontFamily": "undefined, Arial, sans-serif",
      "height": "116px",
      "width": "275px",
    }
  }
>
  <div
    className="playlist-top draggable"
  >
    <div
      className="playlist-top-left draggable"
    />
    <div
      className="playlist-top-title draggable"
    />
    <div
      className="playlist-top-right draggable"
    >
      <div
        id="playlist-shade-button"
        onClick={[Function]}
      />
      <div
        id="playlist-close-button"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    className="playlist-middle draggable"
  >
    <div
      className="playlist-middle-left draggable"
    />
    <div
      className="playlist-middle-center"
    >
      <div
        className="playlist-tracks"
      >
        <div>
          <div
            className="playlist-track current"
            onClick={[Function]}
            onContextMenu={[Function]}
            style={
              Object {
                "backgroundColor": null,
                "color": undefined,
              }
            }
          >
            <div
              className="playlist-track-number"
            >
              1
              .
            </div>
            <div
              className="playlist-track-title"
            >
              <span>
                DJ Mike Llama
                 - 
                Llama Whipping Intro
              </span>
            </div>
            <div
              className="playlist-track-duration"
            >
              03:41
            </div>
          </div>
          <div
            className="playlist-track"
            onClick={[Function]}
            onContextMenu={[Function]}
            style={
              Object {
                "backgroundColor": null,
                "color": null,
              }
            }
          >
            <div
              className="playlist-track-number"
            >
              2
              .
            </div>
            <div
              className="playlist-track-title"
            >
              <span>
                Marilyn Manson
                 - 
                Rock Is Dead
              </span>
            </div>
            <div
              className="playlist-track-duration"
            >
              03:41
            </div>
          </div>
          <div
            className="playlist-track selected"
            onClick={[Function]}
            onContextMenu={[Function]}
            style={
              Object {
                "backgroundColor": undefined,
                "color": null,
              }
            }
          >
            <div
              className="playlist-track-number"
            >
              3
              .
            </div>
            <div
              className="playlist-track-title"
            >
              <span>
                Propellerheads
                 - 
                Spybreak! (Short One)
              </span>
            </div>
            <div
              className="playlist-track-duration"
            >
              02:51
            </div>
          </div>
          <div
            className="playlist-track"
            onClick={[Function]}
            onContextMenu={[Function]}
            style={
              Object {
                "backgroundColor": null,
                "color": null,
              }
            }
          >
            <div
              className="playlist-track-number"
            >
              4
              .
            </div>
            <div
              className="playlist-track-title"
            >
              <span>
                Rob D
                 - 
                Clubbed to Death
              </span>
            </div>
            <div
              className="playlist-track-duration"
            >
              03:35
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="playlist-middle-right draggable"
    >
      <div
        className="rc-slider rc-slider-vertical playlist-scrollbar"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={undefined}
      >
        <div
          className="rc-slider-rail"
          style={Object {}}
        />
        <div
          className="rc-slider-track"
          style={
            Object {
              "bottom": "0%",
              "height": "0%",
              "visibility": "visible",
            }
          }
        />
        <div
          className="rc-slider-step"
        />
        <div
          className="playlist-scrollbar-handle"
        />
        <div
          className="rc-slider-mark"
        />
      </div>
    </div>
  </div>
  <div
    className="playlist-bottom draggable"
  >
    <div
      className="playlist-bottom-left draggable"
    >
      <div
        className="playlist-menu"
        id="playlist-add-menu"
        onClick={[Function]}
      >
        <div
          className="bar"
        />
      </div>
      <div
        className="playlist-menu"
        id="playlist-remove-menu"
        onClick={[Function]}
      >
        <div
          className="bar"
        />
      </div>
      <div
        className="playlist-menu"
        id="playlist-selection-menu"
        onClick={[Function]}
      >
        <div
          className="bar"
        />
      </div>
      <div
        className="playlist-menu"
        id="playlist-misc-menu"
        onClick={[Function]}
      >
        <div
          className="bar"
        />
      </div>
    </div>
    <div
      className="playlist-bottom-center draggable"
    />
    <div
      className="playlist-bottom-right draggable"
    >
      <div
        className="playlist-visualizer"
        onClick={[Function]}
      />
      <div
        className="playlist-running-time-display draggable"
      >
        <div>
          <div
            className=" character character-48"
          >
            0
          </div>
          <div
            className=" character character-50"
          >
            2
          </div>
          <div
            className=" character character-58"
          >
            :
          </div>
          <div
            className=" character character-53"
          >
            5
          </div>
          <div
            className=" character character-49"
          >
            1
          </div>
          <div
            className=" character character-47"
          >
            /
          </div>
          <div
            className=" character character-50"
          >
            2
          </div>
          <div
            className=" character character-55"
          >
            7
          </div>
          <div
            className=" character character-58"
          >
            :
          </div>
          <div
            className=" character character-49"
          >
            1
          </div>
          <div
            className=" character character-56"
          >
            8
          </div>
        </div>
      </div>
      <div
        className="playlist-action-buttons"
      >
        <div
          className="playlist-previous-button"
        />
        <div
          className="playlist-play-button"
          onClick={[Function]}
        />
        <div
          className="playlist-pause-button"
          onClick={[Function]}
        />
        <div
          className="playlist-stop-button"
          onClick={[Function]}
        />
        <div
          className="playlist-next-button"
        />
        <div
          className="playlist-eject-button"
          onClick={[Function]}
        />
      </div>
      <div
        className="playlist-menu"
        id="playlist-list-menu"
        onClick={[Function]}
      >
        <div
          className="bar"
        />
      </div>
      <div
        id="playlist-resize-target"
        onMouseDown={[Function]}
      />
    </div>
  </div>
</div>
`;
